### ISSUE TITLE
[REVIEW]: ESVM: An Electrostatic Vlasov-Maxwell Open-source Code

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3618]
**Submitting author:** @michaeltouati (<a href="http://orcid.org/0000-0001-7590-0941">Michael Touati</a>)
**Repository:** <a href="https://github.com/michaeltouati/ESVM" target ="_blank">https://github.com/michaeltouati/ESVM</a>
**Version:** esvm-v1.0
**Editor:** @danielskatz
**Reviewers:** @tclune, @TomGoffrey, @rouson
**Archive:** <a href="https://doi.org/10.5281/zenodo.5733971" target="_blank">10.5281/zenodo.5733971</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/d0f6a6298710d69b63a8a62d843f8f88/status.svg)](https://joss.theoj.org/papers/d0f6a6298710d69b63a8a62d843f8f88)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/d0f6a6298710d69b63a8a62d843f8f88"><img src="https://joss.theoj.org/papers/d0f6a6298710d69b63a8a62d843f8f88/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/d0f6a6298710d69b63a8a62d843f8f88/status.svg)](https://joss.theoj.org/papers/d0f6a6298710d69b63a8a62d843f8f88)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/michaeltouati/ESVM" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@tclune & @TomGoffrey, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @danielskatz know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @tclune
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/michaeltouati/ESVM">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/michaeltouati/ESVM">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@michaeltouati) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @TomGoffrey
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/michaeltouati/ESVM">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/michaeltouati/ESVM">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@michaeltouati) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @rouson
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/michaeltouati/ESVM">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/michaeltouati/ESVM">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@michaeltouati) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @tclune, @TomGoffrey it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- None

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1039

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=1.33 s (9.0 files/s, 1556.5 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Fortran 90                       7             89            183            946
Python                           2             41             30            573
Markdown                         1             15              0             88
make                             1             31             31             37
TeX                              1              1              0              0
-------------------------------------------------------------------------------
SUM:                            12            177            244           1644
-------------------------------------------------------------------------------


Statistical information for the repository '6523eceeed6206c919c34324' was
gathered on 2021/08/16.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
michaeltouati                    3          1288            644          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Michael Touati              644          100.0          0.0                4.66
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
Note from @tclune (16 Aug 2021):
>I'm a bit swamped at work over the next couple of weeks - if the review can be delayed until early/mid September, you can consider me signed on.

**Comment by danielskatz** [comment]
@tclune and @TomGoffrey - Thanks for agreeing to review this submission.
This is the review thread for the paper. All of our communications will happen here from now on.

Both reviewers have checklists at the top of this thread with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

Please read the first couple of comments in this issue carefully, so that you can accept the invitation from JOSS and be able to check items, and so that you don't get overwhelmed with notifications from other activities in JOSS.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention `openjournals/joss-reviews#3618` so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for reviews to be completed within about 2-4 weeks. Please let me know if either of you require some more time. We can also use Whedon (our bot) to set automatic reminders if you know you'll be away for a known period of time.

Please feel free to ping me (@danielskatz) if you have any questions/concerns.

**Comment by TomGoffrey** [comment]
The code appears to compile and run correctly. However, when running `make extract` I get several pages of warnings:

```
MatplotlibDeprecationWarning: Passing parameters norm and vmin/vmax simultaneously is deprecated since 3.3
and will become an error two minor releases later. Please pass vmin/vmax directly to the norm when creating
it.
```

and

```
MatplotlibDeprecationWarning: shading='flat' when X and Y have the same dimensions as C is deprecated
since 3.3.  Either specify the corners of the quadrilaterals with X and Y, or pass shading='auto',
'nearest' or 'gouraud', or set rcParams['pcolor.shading'].  This will become an error two minor
releases later.
```

These are fairly trivial to fix, and would improve the user experience. They also won't be warnings forever.

I also have a query about the author list (@danielskatz you might want to comment on this?) - The JOSS guidelines state

```
Author lists must be correct and complete. All listed authors must have made a contribution to the work,
and all significant contributors should be included in the author list.
```

It looks like a second user made the initial commit, or at least at some point made a single very large commit. Could you confirm that the current (single) author list conforms to the above guideline?

A note on the paper- the Beam Warming method should be capitalised (after R. F. Warming and R. M. Beam).

I'll continue my review, but wanted to bring these to your attention in the meantime.

**Comment by TomGoffrey** [comment]
There's also a floating point exception when compiled in debug mode, running with the input deck as shipped in the main directory:

```
Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.

Backtrace for this error:
#0  0x7ff518c6b2ed in ???
#1  0x7ff518c6a503 in ???
#2  0x7ff5182d103f in ???
#3  0x7ff518938ea4 in ???
#4  0x7ff5188c11fe in ???
#5  0x55b21ea2f0e5 in __library_MOD_init_simu
	at sources/library.f90:115
#6  0x55b21ea31e78 in esvm
	at sources/esvm.f90:47
#7  0x55b21ea32c8a in main
	at sources/esvm.f90:11
```

**Comment by danielskatz** [comment]
@whedon add @rouson as reviewer

**Comment by whedon** [comment]
OK, @rouson is now a reviewer

**Comment by danielskatz** [comment]
@whedon re-invite @rouson as reviewer

**Comment by whedon** [comment]
The reviewer already has a pending invite.

@rouson please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by danielskatz** [comment]
@rouson - Thanks for agreeing to review this submission.
This is the review thread for the paper. All of our communications will happen here from now on.

You have a checklist at the top of this thread with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

Please read the first couple of comments in this issue carefully, so that you can accept the invitation from JOSS and be able to check items, and so that you don't get overwhelmed with notifications from other activities in JOSS.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention openjournals/joss-reviews#3618 so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for reviews to be completed within about 2-4 weeks. Please let me know if either of you require some more time. We can also use Whedon (our bot) to set automatic reminders if you know you'll be away for a known period of time.

Please feel free to ping me (@danielskatz) if you have any questions/concerns.

**Comment by danielskatz** [comment]
Re the author question from @TomGoffrey 

> I also have a query about the author list (@danielskatz you might want to comment on this?) - The JOSS guidelines state

>>Author lists must be correct and complete. All listed authors must have made a contribution to the work,
and all significant contributors should be included in the author list.

>It looks like a second user made the initial commit, or at least at some point made a single very large commit. Could you confirm that the current (single) author list conforms to the above guideline?

I'm curious to hear what @michaeltouati says before I comment.

**Comment by michaeltouati** [comment]
Toi whom it may concern,

	I am the only one contributor to this version of the code.

My friend, Nitin Shukla, added the openACC instruction for parallelization on GPUs on another version. I've made the mistake to accept its commit without checking it was working for all test cases and I’ve noticed it wasn’t. I thus came back to the original version after that. You can ask him.

It would be a pleasure to add him in the list of authors but he didn’t maker any contribution to this version of the code.

Best regards,
Michaël

**Comment by danielskatz** [comment]
@TomGoffrey - Are you ok with the answer from @michaeltouati?

**Comment by TomGoffrey** [comment]
@michaeltouati Thanks for the clarification.

@danielskatz Fine with with me.

**Comment by whedon** [comment]
:wave: @TomGoffrey, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @tclune, please update us on how your review is going (this is an automated reminder).

**Comment by tclune** [comment]
I hope to start this afternoon.

**Comment by danielskatz** [comment]
👋 @tclune, @TomGoffrey, @rouson - we're now about 3 weeks into the process. How is it going? I see some progress by @tclune and @TomGoffrey, but no checked items from @rouson

**Comment by tclune** [comment]
I was on vacation for the last week.  Last I checked, I was waiting on some response from the submitter on an issue.

**Comment by danielskatz** [comment]
@tclune - can you mention `openjournals/joss-reviews#3618` in your issue so that it (and its status) is visible here?

**Comment by TomGoffrey** [comment]
@danielskatz Apologies, was on vacation last week. Will get started on the review this week.

**Comment by tclune** [comment]
@danielskatz Thank you for the guidance.   I've added the reference to 2 open issues.   (A 3rd issue is just a minor note on quality that will not hold up the review.)

**Comment by TomGoffrey** [comment]
@tclune Whilst it's true that the 3rd issue doesn't directly block any of the review checklists, I'd personally rather see it fixed during the review process. Those warnings will become errors in future versions of matplotlib. Similar for the FPE after compiling in debug mode. Doesn't directly block any of the checklist, but clearly not desirable.

**Comment by danielskatz** [comment]
👋 @michaeltouati - any news on work on the open issues?

**Comment by michaeltouati** [comment]
Dear Daniel,
	
	Sorry for my late reply. I was on vacations and I start today a two day presencial formation. 
I’ll be back at my lab. on Friday and make all the needed corrections / improvements by the end of the week.

I’ve indeed seen and followed that Tom Clune rose some issues that, when I will fixe them, will make the code and the paper much better.

Thanking you,
Best regards,
Michaël

**Comment by danielskatz** [comment]
👋 @michaeltouati - just checking again: any news on work on the open issues?

**Comment by danielskatz** [comment]
👋 @rouson - have you been able to get started on your review?

**Comment by tclune** [comment]
Note - I may be beginning an unplanned, mandatory "vacation" of unknown duration starting on Friday ...

**Comment by danielskatz** [comment]
@tclune - isn't it great to have a job with a high salary and high security :)

**Comment by michaeltouati** [comment]
@danielskatz - If you have such a job for me, I'm in. :).  My current contract indeed ends this December and I'm actively searching for a new job.

@tclune  and @TomGoffrey  - Thank you very much for your review. As you suggested :
- I've made the code Fortran 90 standard compliant
- I've solved the different Matplotlib Deprecation Warning issues
- I've capitalized "Beam-Warming", inline with other eponymous schemes (this is like they chose the numerical scheme they developed in agreement with their family name!)
- I've mentioned the non-linear solvers use a TVD flux limiter
- I moved the perspectives of code additions later in the paper
- I have added in the community guidelines for who this code can be useful (under and post-graduate students, teachers and researchers)
- I've solved the compiling warning error when compiled in debug mode (it was caused by double precision underflows)
- I've added a make test in the makefile (that copy and run pre-prepared input-decks and compared with the corresponding simulation by comparing the terminal output containing the different energy contributions and total energy in each simulation using a diff). I've tested it with gnu and intel compilers and all tests pass if they are compiled with double precision.
- I've added a make check that allows the user to check the simulation parameters he has filled in the input deck 

I'll improve the paper.md tomorrow and I'll try to finish before you go on vacation @tclune . Following your suggestions :
- I'll add add references in the paper.bib of similar codes and Plasma Physics needed literature and papers related with the  academic cases that I provide and that are located in test-cases/
- I'll expand the description of figures by adding the derivation of analytical estimates for the different academic cases that I provide and I'll explain that, by running the same input-deck that I provide, the reader/user should obtain the same figures that I provide. Indeed, I didn't add the simulation data output of the academic cases in order to limit the size of the code to ~100 MB and not GBs (it is a kinetic code!). Also, I'll add a discussion about the velocity bin sizes there.
- Finally, I'll add in the readme and in the paper.md a description of the make check and make test that I have added.

Again, thank you very much for your suggestions / corrections. I think the code and the paper will be greatly improved thanks to you. I have been happy to learn it is already currently used by a PhD student who found the code thanks to GitHub and use it for his PhD investigations in Theoretical Plasma Physics.

Best regards,
Michaël

**Comment by rouson** [comment]
@danielskatz I was just getting started and noticed that all of the checkboxes on my list appear to be grayed out.  I also see that I'm not listed under "Assigned," whereas the other two reviewers are listed so I assume you have to assign this issue to me for me to review it.

**Comment by danielskatz** [comment]
@rouson - you have to accept the invitation to review, as mentioned early in this issue.  It's probably expired, so I'll re-issue it, which may also say it's expired when you click on it (a strange github bug) - if so, let me know

**Comment by danielskatz** [comment]
@whedon re-invite @rouson as reviewer

**Comment by whedon** [comment]
The reviewer already has a pending invite.

@rouson please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by danielskatz** [comment]
@whedon re-invite @rouson as reviewer

**Comment by whedon** [comment]
The reviewer already has a pending invite.

@rouson please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by michaeltouati** [comment]
Dear all, 
it seems to me I finished today to improve the code and the paper in order to answer all very useful corrections / suggestions / additions of @tclune and @TomGoffrey. 
I hope it will be sufficient.

**Comment by michaeltouati** [comment]
@whedon generate pdf

**Comment by danielskatz** [comment]
👋 @rouson - we're now about 7 weeks into the process. How is it going?

**Comment by michaeltouati** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: :point_left:

**Comment by rouson** [comment]
@danielskatz the most recent invitation has expired.

**Comment by danielskatz** [comment]
@whedon re-invite @rouson as reviewer

**Comment by rouson** [comment]
@danielskatz thanks for the new invitation.  I just accepted it.   My initial thoughts are below.  I'll submit these as issues on the repository and will offer pull requests in some cases.  Let me know if I should delete the text below once I've added the corresponding issues on pull requests on the submitted repository.

1. In a [comment] above, @michaeltouati mentions having "made the code Fortran 90 standard compliant" and the repository "About" text describes it as Fortran 95.  Newer Fortran standards are very nearly backward compatible, but do declare some older features obsolescent.  Unless there is a specific reason to limit oneself to Fortran 95 (e.g., if it uses features that were deleted in later standards), I recommend simply describing it as "standard-compliant Fortran."  Better yet, I recommend evaluating whether Fortran 2003, 2008, and 2018 offer anything of value.  A lot has happened in the language since 1990-'95.
2.  I highly recommend evaluating the Fortran Package Manager ([`fpm`]) first released in 2020.  At least for purposes of building, it's possible to replace the 415-line `makefile` with a 4-line `fpm` manifest in less than 5 minutes as I just verified by doing so while running a stopwatch.  I'll post a link to a fork that demonstrates this approach.  In conducting my review, I'll take a look at what additional capabilities the `makefile` and whether `fpm` offers the same -- hopefully at greatly reduced complexity in terms of lines of build-system code.   For now, I can say that `fpm` can also run tests and examples without a lot of the repetitive code in the `makefile`, which appears to be poised to grow proportionately with the number of tests and the growth will comprise making repeated copies of `makefile` instructions.
3. The repository takes surprisingly long to clone, probably owing to the large number of `.png` files.  I strongly recommend against storing this many generated binary files in the repository unless for some reason they are required for execution and/or testing.   Binary files make it difficult or impossible for `git` to optimize space by storing diffs.  Changing one character might mean that `git` has to replace the entire file.  Worse still, the old versions of binary files stay in the repository forever.   The only way to remove them is to revise history, which is a discouraged practice.  I'm not sure of the best solution, but there are tools that perform the necessary repo surgery.  See "[Don't ever commit binary files to git]."
4. Explicit deallocation of `allocatable` entities is unnecessary unless either the compiler is broken or one there is specific reason to do so and that reason is backed up with data on memory system performance.  As my book co-author Jim Xia used to day, "Let the compiler do it's job."
5. I usually recommend against non-default array bounds because there are several scenarios in which the bounds are not preserved.  I have some [demonstration code] online that generates a table of some such scenarios.   If a code uses non-default bounds, it might be good to have a  style guide that prescribes behaviors that will avoid pitfalls.  Preferably such a guide would be enforced by continuous-integration tests.

After putting the above comments in issues and pull requests, I'll add more thoughts.


[`fpm`]: https://github.com/fortran-lang/fpm
[comment]: https://github.com/openjournals/joss-reviews/issues/3618#issuecomment-930530903
[demonstration code]: https://github.com/sourceryinstitute/fidbits/tree/master/src/array-bounds-preservation
[Don't ever commit binary files to git]: https://robinwinslow.uk/dont-ever-commit-binary-files-to-git

**Comment by michaeltouati** [comment]
Dear Damian @rouson,
Thank you very much for your first initial thoughts :
1. I just recently switch to 95 because I wanted to use the intrinsic cpu_time fortran timer that were not Fortran 90 standard compliant
2. I'll investigate about fpm. It is true that the makefile can become very large if the number of tests further increase. One simpler solution would be to delete some tests or try to make a loop. 
3. I have suppressed all the png files from simulation examples except the ones needed to compile the paper.md Should I suppress all the commit history now? What would be the best way to do it according to you? If I do it, the cloning would become almost instantaneous (all files would represents only few MB)
4. I'll follow your recommendation by deleting the deallocation instructions. It's just that I was thinking the opposite... the more pieces of information one gives to the compiler, the best it is.
5. I needed to declare arrays with non-default array bounds to be able to change the simulation parameters in the input-deck without having to recompile the code, the bounds of the arrays depending on the choice of the user written in the input-deck. I'll have a look at your demonstration code tomorrow and I'll add in the contributing.md file a style guide to avoid such problems.

**Comment by michaeltouati** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
👋 @rouson - any news?  Did you manage to do this

> I still need to move my comments to issues on the repository and create a pull request with suggestions. I hope to have time for this this evening or at the latest would be this weekend.

Please remember to add `openjournals/joss-reviews#3618` in the issues/PRs you open, so that they are visible here, and whether or not they are open or closed is also visible here.

**Comment by danielskatz** [comment]
👋 @TomGoffrey - this review seems to be dragging. Perhaps you can all update us on where you are, and what is needed for you to make progress (check off items on your list)?

**Comment by danielskatz** [comment]
I've just emailed @TomGoffrey and @rouson

**Comment by TomGoffrey** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: :point_left:

**Comment by TomGoffrey** [comment]
@danielskatz Apologies for the covid and teaching induced delay. I'm happy to sign off on this, @michaeltouati has done an excellent job addressing my original concerns.

**Comment by danielskatz** [comment]
Thanks @TomGoffrey

**Comment by danielskatz** [comment]
👋  @rouson - if you still want to review this, please let me know ASAP.  If not, since you were not an original reviewer, and I have two completed reviews, I will go ahead and move forward on accepting it.

**Comment by danielskatz** [comment]
👋 @rouson - I'm going to go ahead with accepting this now.  Please feel free to make any comments to the author in the source repo, but we won't consider that part of the review, which will now end.

**Comment by danielskatz** [comment]
👋 @michaeltouati - I'm now going to proofread the paper

**Comment by danielskatz** [comment]
👋 @michaeltouati - I've suggested some changes in https://github.com/michaeltouati/ESVM/pull/15.  In addition, this paper is too long.  See https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain.  Specifically, please remove the sections: 
* Equations computed by ESVM
* ESVM units
* ESVM numerical stability
* ESVM Plasma Physics academic case simulations
* Any corresponding figures

This material could perhaps go in one or more supplementary file(s) in the repository (which you could point to in the JOSS paper).

In addition, please check the references to make sure all journals and books are in Title Case, and are not acronyms (e.g., JETP). This can be done by protecting {L}etters or {Words} in the bibtex entry. Finally, just as a note, in English, there is never a space before a ":".

**Comment by michaeltouati** [comment]
I have accepted your pull request @danielskatz .

Also, following your suggestion, I removed the mentioned sections in the JOSS paper and I re-wrote them in separate pdf files in the code repo. 

Finally, I have checked and corrected the references to make sure they are all in Title Case.

**Comment by michaeltouati** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1103/PhysRev.94.511 is OK
- 10.1103/PhysRev.107.1 is OK
- 10.1103/PhysRev.108.546 is OK
- 10.1063/1.1706638 is OK
- 10.1109/TAP.1966.1138693 is OK
- 10.1103/PhysRevLett.19.219 is OK
- 10.1146/annurev.fl.18.010186.002005 is OK
- 10.1103/PhysRevLett.59.1817 is OK
- 10.1088/1367-2630/16/7/073014 is OK
- 10.21105/joss.02182 is OK

MISSING DOIs

- None

INVALID DOIs

- https://doi.org/10.1007/BF01448839 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1002/cpa.3160050303 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1002/cpa.3160130205 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1016/0021-9991(68)90015-6 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1016/0021-9991(76)90110-8 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1016/0021-9991(77)90094-8 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1016/0021-9991(79)90145-1 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1006/jcph.1994.1159 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1006/jcph.1994.1187 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1016/j.jcp.2004.06.007 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1016/j.jcp.2009.04.005 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1016/j.jcp.2011.04.034 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1016/j.cpc.2014.03.004 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1016/j.cpc.2017.09.024 is INVALID because of 'https://doi.org/' prefix
```

**Comment by danielskatz** [comment]
@michaeltouati - I've created one more PR to try to fix two lists: https://github.com/michaeltouati/ESVM/pull/16
In addition, please remove the 'https://doi.org/' prefix from the DOIs listed above.

**Comment by michaeltouati** [comment]
@danielskatz - I've accepted your pull request and removed all "https://doi.org/" from the paper.bib file.

**Comment by danielskatz** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/BF01448839 is OK
- 10.1002/cpa.3160050303 is OK
- 10.1103/PhysRev.94.511 is OK
- 10.1103/PhysRev.107.1 is OK
- 10.1103/PhysRev.108.546 is OK
- 10.1002/cpa.3160130205 is OK
- 10.1063/1.1706638 is OK
- 10.1109/TAP.1966.1138693 is OK
- 10.1103/PhysRevLett.19.219 is OK
- 10.1016/0021-9991(68)90015-6 is OK
- 10.1016/0021-9991(76)90110-8 is OK
- 10.1016/0021-9991(77)90094-8 is OK
- 10.1016/0021-9991(79)90145-1 is OK
- 10.1146/annurev.fl.18.010186.002005 is OK
- 10.1103/PhysRevLett.59.1817 is OK
- 10.1006/jcph.1994.1159 is OK
- 10.1006/jcph.1994.1187 is OK
- 10.1016/j.jcp.2004.06.007 is OK
- 10.1016/j.jcp.2009.04.005 is OK
- 10.1016/j.jcp.2011.04.034 is OK
- 10.1088/1367-2630/16/7/073014 is OK
- 10.1016/j.cpc.2014.03.004 is OK
- 10.1016/j.cpc.2017.09.024 is OK
- 10.21105/joss.02182 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03618/joss.03618/10.21105.joss.03618.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
Just one more change, I think - sorry I didn't see this before.
https://github.com/michaeltouati/ESVM/pull/17

**Comment by danielskatz** [comment]
Once you've accepted this, can you

- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata. This includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it). You may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

I can then move forward with accepting the submission.

**Comment by michaeltouati** [comment]
@danielskatz - done
https://zenodo.org/record/5720926#.YZz1bXvMJH4 (DOI : 10.21105/joss.03618)

**Comment by danielskatz** [comment]
@michaeltouati - unfortunately, you've given the JOSS DOI (for the paper) to zenodo. Here, we need to the DOI of the archived software itself. The JOSS DOI will be for the paper. Let zenodo create its own DOI for the software, and give us that.

And I think the version tag is "esvm-v1.0" - is this correct?

**Comment by michaeltouati** [comment]
Yes, "esvm-v1.0" is correct @danielskatz 

I've also modified the DOI. it is now:
10.5282/zenodo.5720926

**Comment by danielskatz** [comment]
@whedon set esvm-v1.0 as version

**Comment by whedon** [comment]
OK. esvm-v1.0 is the version.

**Comment by danielskatz** [comment]
@whedon set 10.5282/zenodo.5720926 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5282/zenodo.5720926" target="_blank">10.5282/zenodo.5720926</a> is the archive.

**Comment by danielskatz** [comment]
@michaeltouati - unfortunately, that DOI does not resolve currently.  I think something got messed up in the process.  I suggest you either contact zenodo help and see if they can fix the problem, or create a new archive where the doi is automatically generated, the contact zenodo help to ask them to remove the old one.

**Comment by michaeltouati** [comment]
I've send a message to the zenodo help contact. I'm waiting for their answer.

**Comment by michaeltouati** [comment]
@danielskatz - due to the absence of answer to my message from the zenodo help contact, I have uploaded another time the software. Here is the DOI :
https://doi.org/10.5281/zenodo.5733971

**Comment by danielskatz** [comment]
@whedon set 10.5281/zenodo.5733971 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.5733971" target="_blank">10.5281/zenodo.5733971</a> is the archive.

**Comment by danielskatz** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/BF01448839 is OK
- 10.1002/cpa.3160050303 is OK
- 10.1103/PhysRev.94.511 is OK
- 10.1103/PhysRev.107.1 is OK
- 10.1103/PhysRev.108.546 is OK
- 10.1002/cpa.3160130205 is OK
- 10.1063/1.1706638 is OK
- 10.1109/TAP.1966.1138693 is OK
- 10.1103/PhysRevLett.19.219 is OK
- 10.1016/0021-9991(68)90015-6 is OK
- 10.1016/0021-9991(76)90110-8 is OK
- 10.1016/0021-9991(77)90094-8 is OK
- 10.1016/0021-9991(79)90145-1 is OK
- 10.1146/annurev.fl.18.010186.002005 is OK
- 10.1103/PhysRevLett.59.1817 is OK
- 10.1006/jcph.1994.1159 is OK
- 10.1006/jcph.1994.1187 is OK
- 10.1016/j.jcp.2004.06.007 is OK
- 10.1016/j.jcp.2009.04.005 is OK
- 10.1016/j.jcp.2011.04.034 is OK
- 10.1088/1367-2630/16/7/073014 is OK
- 10.1016/j.cpc.2014.03.004 is OK
- 10.1016/j.cpc.2017.09.024 is OK
- 10.21105/joss.02182 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2772

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2772, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1465310795491954697) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2773
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03618](https://doi.org/10.21105/joss.03618)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by danielskatz** [comment]
Congratulations @michaeltouati (Michael Touati)!!

And thanks to @tclune and @TomGoffrey for reviewing!
We couldn't do this without you

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03618/status.svg)](https://doi.org/10.21105/joss.03618)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03618">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03618/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03618/status.svg
   :target: https://doi.org/10.21105/joss.03618
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03618/status.svg)](https://doi.org/10.21105/joss.03618)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
